### PR TITLE
Add Streams Implementation

### PR DESCRIPTION
**Pull request by nafraf**
Hi, this PR is to implement the Stream functions which return an scalar. 
This is related to the issue #39 
Comments are welcome.

<details>
<summary>
This is the sample output of the test file:  <b>Streams.ino</b>
</summary>

```sh
Connecting to the WiFi.
IP Address: 192.168.1.6
Connected to the Redis server!
XADD mystream * name Sarah
1687655985023-0
XADD mystream * name John
1687655985028-0
XLEN mystream
4
XDEL mystream 1687655985023-0
1
XLEN mystream
3
XGROUP CREATE mystream mygroup $ MKSTREAM
1
XGROUP CREATECONSUMER mystream mygroup myconsumer
1
XGROUP DESTROY mystream mygroup 
1
Connection closed!
```

</details>

To implement the remaining functions, I still need to solve how to return Arrays with Arrays inside. I'll appreciate if you have some ideas. Thanks.

### CODE REVIEW

**Comment by rpj** [comment]
@nafraf This is absolutely excellent work, thank you so much for your contribution!

> To implement the remaining functions, I still need to solve how to return Arrays with Arrays inside. I'll appreciate if you have some ideas. Thanks.

I think the simplest way - and the one that fits with the library's current preference of utilizing C++ standard library container classes whenever possible - would be to simply return nested `std::vector`s. I'm up for other suggestions, but off the top of my head this one makes the most sense overall.

File: Redis.cpp
Comment (rpj) [inline Redis.cpp:221]: Given that the values for `exact` are a (small) bounded set, I think I'd prefer to see them expressed as an `enum`.

**Comment by rpj** [comment]
I'll hold off on a full review until you've had a chance to finish the nested-array implementation @nafraf, but for what it's worth I did read over the current contents of this PR and they look _exceptionally good_. I had only one comment, but otherwise the structure, style and commentary is exactly what I want and expect of a PR. I can already tell you that this one will be approved easily when you're ready! 😄 👍

**Comment by rpj** [comment]
I've linked this PR to https://github.com/electric-sheep-co/arduino-redis/issues/39 because it will indeed resolve that issue once merged. 👍 🚀

**Comment by rpj** [comment]
@holla2040 yup you got it, that's right! Also please don't feel obligated to review: I simply added you as you've been extremely helpful in the past as the unofficial co-maintainer* and so in the interest of transparency and information dissemination, I thought I'd add you to all future PRs so that you're at least aware of them. Thank you again! 

\* Let's make it official! Congratulations and thank you for your service 😄 🙇  👍

**Comment by holla2040** [comment]
No, I'll take a look. I've had numerous projects with time series data, just haven't used redis streams for this before. I'll act as the end user to test the example code.

**Comment by rpj** [comment]
Excellent, you're awesome @holla2040, thank you!! 🙌

**Comment by holla2040** [comment]
OK, please keep in mind that I haven't used redis streams before but I do understand their use case. I ran the streams example, it compiled and ran. Couple thoughts.

1. I manually entered all the printed test_xgroup redis X-commands into redis-cli.

hollabaugh.com:6379> XRANGE mystream - +
1) 1) "1688668875026-0"
   2) 1) "name"
      2) "Sarah"
2) 1) "1688668877169-0"
   2) 1) "name"
      2) "John"
3) 1) "1688668881233-0"
   2) 1) "name"
      2) "Ginger"
4) 1) "1688668883410-0"
   2) 1) "name"
      2) "Kyle"
hollabaugh.com:6379> XGROUP CREATE mystream mygroup $ MKSTREAM
OK
hollabaugh.com:6379> XGROUP CREATECONSUMER mystream mygroup myconsumer
(error) ERR Unknown subcommand or wrong number of arguments for 'CREATECONSUMER'. Try XGROUP HELP.
hollabaugh.com:6379> XINFO CONSUMERS mystream mygroup
(empty array)

Streams.ino line 103   sprintf(charBuf, "XGROUP CREATECONSUMER %s %s %s", STREAMS_KEY, STREAMS_GROUP, STREAMS_CONSUMER);
 has a typo or something.

2. XREAD def in Redis.h doesn't support blocking. https://redis.io/commands/xread/. This might be of limited use but probably my inexperience showing here.

3. The example is more a test harness and than example for users. It does show streams functionality, so its probably fine.
4.

**Comment by nafraf** [comment]
@holla2040 Thanks.
1. ``XREAD``: The implementation was incomplete. It was fixed on [464efd7](https://github.com/electric-sheep-co/arduino-redis/pull/64/commits/464efd7750137e80a61e9817e5b756ef1aa962f0).
2. ``XGROUP CREATECONSUMER``: Which version of Redis Server are you using? That command is available since version 6.2.0

**Comment by holla2040** [comment]
redis_version:6.0.16 on ubuntu 22.04. I'll update it.

**Comment by holla2040** [comment]
I'm running redis on a production AWS instance running stock ubuntu 22.04. updating to 6.2 wasn't available through standard package manager. If you have a PPA for 6.2, that will work. Running 7.x broke redis, so I reverted back to 6.0.x

**Comment by nafraf** [comment]
Maybe this [PPA](https://launchpad.net/~redislabs/+archive/ubuntu/redis)?

**Comment by holla2040** [comment]
using that PPA upgraded my install to 7.0.12. I'll check out my comment from above.

File: Redis.cpp
Comment (rpj) [inline Redis.cpp:106]: Great thought adding handling for this and my bad for not doing so initially! Thank you! 🙇 

However, I worry a bit that this could become an issue for consumers that aren't aware of the implementation and don't explicitly check the first element in the returned vector for an error string, because otherwise it will appear to them as a valid one-element list.

However, I'm realizing now that the interface of this method currently doesn't allow passing out-of-band error information to the consumer _besides_ raising an exception. Fortunately this is, I believe, a valid place to raise an exception as it would be a programming error.

So I think I'd prefer this raise an exception: what do you think @nafraf, @holla2040?

File: RedisInternal.cpp
Comment (rpj) [inline RedisInternal.cpp:66]: You implemented nested handling here, too?! Huge bonus implementation, you rock @nafraf thank you!! 🚀 🙇

**Review by rpj — COMMENTED** [review]
Overall, I have to say: you outdid yourself @nafraf! My immense thanks for the care and attention to detail you put into this PR: fixing some long-standing bugs, improving error handling, immensely improving the array `String` operator by adding nested handling... all above and beyond what is required here, and all extremely appreciated.

I do have one concern regarding the pattern of "overloading" the semantics of the return value in the error cases: those that follow the pattern that starts with the guard `if(rv->type() == RedisObject::Type::InternalError)`. I have no objection to the implementation: in fact it was a _mistake_ of mine originally to not handle that case, so it is again much appreciated.

My worry is that consumers will have no way to differentiate the error case from a normal one-element return without parsing the first element's string value, which is not acceptable as it is incredibly fragile. Given the way the libraries interfaces are already defined - and accordingly I'm very hesitant to change them - I think our only option in these cases is to raise an exception. Would love to hear your thoughts on this, but I do believe this needs to be fixed one way or another before this PR can land.

Finally, my _sincerest_ apologies for my delay in getting to this review. I meant to get to it all the way back last weekend but real life had other designs. No excuse, just apology. I respect and appreciate the work you've put in here a ton, regardless of my inability to get to this in a timely manner. 😉 😊 

Let me know what you think on the error handling and let's get this one landed! 🛬

**Comment by rpj** [comment]
> My worry is that consumers will have no way to differentiate the error case from a normal one-element return without parsing the first element's string value, which is not acceptable as it is incredibly fragile. Given the way the library's interfaces are already defined - and accordingly I'm very hesitant to change them - I think our only option in these cases is to raise an exception. Would love to hear your thoughts on this, but I do believe this needs to be fixed one way or another before this PR can land.

_Arghhhh_, I'd completely forgotten that exceptions are disabled by default in the Arduino runtime because of the expense associated with them. So exceptions are out. 😦 

Back to the drawing board on this one then, and I think we have no choice but to change method interfaces (necessitating a major version bump when this lands and is released). 

But that does mean that we have a lot of leeway in the implementation path here. A simple implementation could be to instead start returning shared pointers of heap-allocated objects from these methods, with the `nullptr` return indicating error. Trouble with this is a lack of ability to return to the consumer _the actual error information_. 

So another option could be to subclass `std::vector` into one that can optionally contain error information and therefore signal to consumers if the call failed, and then return that from all of these methods instead of a bare `std::vector`.

Would love to hear any other thoughts on this from you guys!

Sorry that this is adding extra work to get this landed @nafraf  but these error-handling changes are genuinely valuable to the library as a whole and accordingly I want to make sure we get them right. Thank you!

**Comment by nafraf** [comment]
Hi @rpj, thanks for the review.
I think the subclass is a good option. It is possible to add a ``string`` to return the error message if it exists.

Honestly, I prefer a simple solution like the [RESP protocol](https://redis.io/docs/reference/protocol-spec/) does, returning a ``-`` as first byte in case of error, but you are right, a solution like that can be confusing for the users.

**Comment by rpj** [comment]
@nafraf 

> Honestly, I prefer a simple solution like the [RESP protocol](https://redis.io/docs/reference/protocol-spec/) does, returning a - as first byte in case of error, but you are right, a solution like that can be confusing for the users.

Hmmm this is a very good point: I do prefer simple solutions as well, and would prefer the simplest solution possible here, too.

To that end, the method you're currently using - with a first-byte of "-" to indicate error per the [RESP protocol](https://redis.io/docs/reference/protocol-spec/) - could be made to work _if_ we give consumers a simple, foolproof way to check for error return values. So something like a small utility function:

```c++
bool isErrorReturn(std::vector<string>& returnVec) {
  // check returnVec's first element for a first byte of "-":
  // return true if yes, false otherwise
}
```

This way we don't require _every_ consumer to do the check themselves every time, and if we _do_ decide to change this interface in the future we could change only the logic in this method and it would be otherwise opaque to consumers.

I'd be perfectly happy with this pattern too, so it's up to you. Thank you!!

**Comment by nafraf** [comment]
@rpj 
Sorry for the delay.
The auxiliary function was included in [2920cc8](https://github.com/electric-sheep-co/arduino-redis/pull/64/commits/2920cc85efd5e329633fbed27a87415ccc44cbd6), but it returns ``bool``. An example of use was added in Streams.ino: ``VALIDATE_VECTOR_STRING_RESULT(result, command)``
Thanks for your comments, if you think that any other change is required, please tell me.

**Review by rpj — APPROVED** [review]
Fantastic work all around, my deepest appreciation again @nafraf. 🙇 

I like the `bool` return from `isErrorReturn()`, it keeps it even simpler, which is always good. Great choice!

LGTM, approved! 👍 Merge whenever you're ready, and again: thank you & welcome to the team officially!

**Comment by rpj** [comment]
@nafraf By the way, for reference there is never any need to apologize for any delay. It's an open source project and I perfectly well understand that working on it is a hobby/side-time thing. It is for me too! So I genuinely mean it when I say I appreciate the effort, because I do understand what it takes to contribute. Always take all the time you want and/or need, anything is appreciated and valuable 😄 🙇

**Comment by nafraf** [comment]
@rpj  The changes were merged. Thanks you.

**Comment by rpj** [comment]
Awesome @nafraf! I'll get the documentation updated and release v3.0 pushed out ASAP! In past experience it can take the Arduino release bot up to a day before it re-indexes the package. So within a day or so of the release being available here it will become available in the Arduino Package Manager too. 🚀

**Pull request merged**
